### ISSUE TITLE
Building docker image fails when compiling LLVM

### ISSUE THREAD

**Issue by CarlCurist** [RUB-SysSec/loki#1]
Hello. I very interested in Loki obfuscator. But when I try to build the docker image, it fails in the process of compiling LLVM. Here is part of the log:

> #12 42.02 [126/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Process.cpp.o
#12 42.70 [127/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Host.cpp.o
#12 42.94 [128/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/YAMLParser.cpp.o
#12 43.67 [129/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/StringMatcher.cpp.o
#12 43.84 [130/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Signals.cpp.o
#12 44.71 [131/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/Path.cpp.o
#12 44.98 [132/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/Error.cpp.o
#12 47.63 [133/2619] Building CXX object lib/Support/CMakeFiles/LLVMSupport.dir/VirtualFileSystem.cpp.o
#12 48.22 [134/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/TGLexer.cpp.o
#12 48.32 [135/2619] Linking CXX static library lib/libLLVMSupport.a
#12 51.77 [136/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/Main.cpp.o
#12 52.20 [137/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/Attributes.cpp.o
#12 183.5 [138/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmMatcherEmitter.cpp.o
#12 183.5 FAILED: utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmMatcherEmitter.cpp.o
#12 183.5 /usr/bin/c++  -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Iutils/TableGen -I/llvm_src/llvm-project/llvm/utils/TableGen -Iinclude -I/llvm_src/llvm-project/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++1y -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG    -fno-exceptions -fno-rtti -MD -MT utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmMatcherEmitter.cpp.o -MF utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmMatcherEmitter.cpp.o.d -o utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmMatcherEmitter.cpp.o -c /llvm_src/llvm-project/llvm/utils/TableGen/AsmMatcherEmitter.cpp
#12 183.5 c++: internal compiler error: Killed (program cc1plus)
#12 183.5 Please submit a full bug report,
#12 183.5 with preprocessed source if appropriate.
#12 183.5 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#12 184.3 [139/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/JSONBackend.cpp.o
#12 207.0 [140/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterEmitter.cpp.o
#12 207.0 FAILED: utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterEmitter.cpp.o
#12 207.0 /usr/bin/c++  -DGTEST_HAS_RTTI=0 -D_GNU_SOURCE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -Iutils/TableGen -I/llvm_src/llvm-project/llvm/utils/TableGen -Iinclude -I/llvm_src/llvm-project/llvm/include -fPIC -fvisibility-inlines-hidden -Werror=date-time -std=c++1y -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wno-missing-field-initializers -pedantic -Wno-long-long -Wimplicit-fallthrough -Wno-maybe-uninitialized -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wno-comment -fdiagnostics-color -ffunction-sections -fdata-sections -O3 -DNDEBUG    -fno-exceptions -fno-rtti -MD -MT utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterEmitter.cpp.o -MF utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterEmitter.cpp.o.d -o utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterEmitter.cpp.o -c /llvm_src/llvm-project/llvm/utils/TableGen/AsmWriterEmitter.cpp
#12 207.0 c++: internal compiler error: Killed (program cc1plus)
#12 207.0 Please submit a full bug report,
#12 207.0 with preprocessed source if appropriate.
#12 207.0 See <file:///usr/share/doc/gcc-7/README.Bugs> for instructions.
#12 207.3 [141/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/AsmWriterInst.cpp.o
#12 208.9 [142/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/CallingConvEmitter.cpp.o
#12 209.4 [143/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/CodeGenHwModes.cpp.o
#12 210.9 [144/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/TGParser.cpp.o
#12 211.3 [145/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/CodeEmitterGen.cpp.o
#12 211.8 [146/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/Record.cpp.o
#12 212.8 [147/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/CodeGenInstruction.cpp.o
#12 219.3 [148/2619] Building CXX object utils/TableGen/CMakeFiles/llvm-tblgen.dir/CodeGenDAGPatterns.cpp.o
#12 220.5 [149/2619] Building CXX object lib/TableGen/CMakeFiles/LLVMTableGen.dir/SetTheory.cpp.o
#12 220.5 ninja: build stopped: subcommand failed.
executor failed running [/bin/sh -c cmake   -G "Ninja"   -DCMAKE_EXPORT_COMPILE_COMMANDS=ON   -DCMAKE_BUILD_TYPE=Release   -DCMAKE_INSTALL_PREFIX=$LLVM_INSTALL_DIR   -DLLVM_CCACHE_DIR=$LLVM_DIR/llvm-project/llvm/ccache   -DLLVM_ENABLE_CXX1Y=On   -DLLVM_ENABLE_IDE=On   -DLLVM_ENABLE_PROJECTS=clang   -DLLVM_TARGETS_TO_BUILD="X86"   -DLLVM_PARALLEL_COMPILE_JOBS=$PARALLEL_JOBS   -DLLVM_PARALLEL_LINK_JOBS=$PARALLEL_JOBS   -DLLVM_USE_LINKER=gold   -DLLVM_BUILD_LLVM_DYLIB=On   /$LLVM_DIR/llvm-project/llvm   && ninja -j ${PARALLEL_JOBS}]: exit code: 1

I tried many times but got the same result.
By the way, could you please provide a docker image in a tar archive? So that I can directly load the image from it. Otherwise I had to spend a lot of time compiling LLVM because clone LLVM's repository in GitHub is too slow in China. 
Thank you very much.

**Comment by mu00d8** [comment]
Hi there,

I cannot reproduce this problem. It builds flawlessly for me on all tested machines. 

Anyway, I've uploaded the image to Dockerhub [https://hub.docker.com/r/mu00d8/loki-obfuscation](https://hub.docker.com/r/mu00d8/loki-obfuscation). After pulling, the image tag is different than the convenience scripts `docker_X.sh` in this repository expect, so you need to change the tag. To save you time and trouble, I've added a convenience script `pull.sh` which will pull the image from Dockerhub and change the tag. Everything should then work as expected.

TL;DR: Run `git pull && ./pull.sh` and you should be good to go.

**Comment by CarlCurist** [comment]
Thank you. It helps me a lot.
